modules/spring/src/main/java/org/apache/ignite/transactions/spring/AbstractSpringTransactionManager.java

```diff
@@ -0,0 +1,303 @@
+                long timeout = dfltTxTimeout;
+
+                if (definition.getTimeout() > 0)
+                    timeout = TimeUnit.SECONDS.toMillis(definition.getTimeout());
+
+                TransactionProxy newTx = txFactory.txStart(transactionConcurrency,
+                    convertToIgniteIsolationLevel(definition.getIsolationLevel()), timeout);
+
+                if (log.isDebugEnabled())
+                    log.debug("Started Ignite transaction: " + newTx);
+
+                txObj.setTransactionHolder(new IgniteTransactionHolder(newTx), true);
+            }
+
+            txObj.getTransactionHolder().setSynchronizedWithTransaction(true);
+            txObj.getTransactionHolder().setTransactionActive(true);
+
+            tx = txObj.getTransactionHolder().getTransaction();
+
+            // Bind the session holder to the thread.
+            if (txObj.isNewTransactionHolder())
+                TransactionSynchronizationManager.bindResource(txFactory, txObj.getTransactionHolder());
+        }
+        catch (Exception ex) {
+            if (tx != null)
+                tx.close();
+
+            throw new CannotCreateTransactionException("Could not create Ignite transaction", ex);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void doCommit(DefaultTransactionStatus status) throws TransactionException {
+        IgniteTransactionObject txObj = (IgniteTransactionObject)status.getTransaction();
+        TransactionProxy tx = txObj.getTransactionHolder().getTransaction();
+
+        if (status.isDebug() && log.isDebugEnabled())
+            log.debug("Committing Ignite transaction: " + tx);
+
+        try {
+            tx.commit();
+        }
+        catch (Exception e) {
+            throw new TransactionSystemException("Could not commit Ignite transaction", e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void doRollback(DefaultTransactionStatus status) throws TransactionException {
+        IgniteTransactionObject txObj = (IgniteTransactionObject)status.getTransaction();
+        TransactionProxy tx = txObj.getTransactionHolder().getTransaction();
+
+        if (status.isDebug() && log.isDebugEnabled())
+            log.debug("Rolling back Ignite transaction: " + tx);
+
+        try {
+            tx.rollback();
+        }
+        catch (Exception e) {
+            throw new TransactionSystemException("Could not rollback Ignite transaction", e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void doSetRollbackOnly(DefaultTransactionStatus status) throws TransactionException {
+        IgniteTransactionObject txObj = (IgniteTransactionObject)status.getTransaction();
+        TransactionProxy tx = txObj.getTransactionHolder().getTransaction();
+
+        assert tx != null;
+
+        if (status.isDebug() && log.isDebugEnabled())
+            log.debug("Setting Ignite transaction rollback-only: " + tx);
```

Comment:
Is `tx` have readable `toString` implementation?

Reply:
Done. But ClientTransaction implementation don't have toSring() implementation itself. I'l file the corresponding ticket.